### PR TITLE
Update RC training components

### PR DESCRIPTION
**Pull request by AnnaKwa**
This modifies existing reservoir model components to work with the cubedsphere training function (followup draft PR is [here](https://github.com/ai2cm/fv3net/pull/2179) for context). 

Sorry the changes are kind of miscellaneous and a pain to go through. I tried to add comments describing the changes in each module to make this clearer.

Coverage reports (updated automatically):

### CODE REVIEW

File: external/fv3fit/fv3fit/reservoir/config.py [hunk 1]
```diff
@@ -59,47 +59,46 @@ class BatchLinearRegressorHyperparameters:
...
-        if self.subdomain is not None:
-            if (
-                self.subdomain.size + 2 * self.subdomain.overlap
-            ) != self.reservoir_hyperparameters.input_size:
-                raise ValueError(
-                    "Subdomain size + overlaps and reservoir input_size must match."
-                )
+    @property
+    def variables(self) -> Set[str]:
+        return set(self.input_variables).union(self.output_variables)
```
Comment (spencerkclark) [inline external/fv3fit/fv3fit/reservoir/config.py:101]: Is this necessary if `output_variables` is a subset of `input_variables`, as stated in the docstring?

File: external/fv3fit/fv3fit/reservoir/config.py [hunk 2]
```diff
@@ -59,47 +59,46 @@ class BatchLinearRegressorHyperparameters:
 
 
 @dataclass
-class ReservoirTrainingConfig:
+class ReservoirTrainingConfig(Hyperparameters):
     """
+    input_variables: variables and additional features in time series
+    output_variables: time series variables, must be subset of input_variables
```
Comment (spencerkclark) [inline external/fv3fit/fv3fit/reservoir/config.py:65]: Should the subset condition be validated?

File: external/fv3fit/tests/reservoir/test_domain.py [hunk 3]
```diff
@@ -129,17 +149,89 @@ def test_stack_time_series_samples():
...
+        input_with_halo, with_overlap=True
+    )
+
+    # 4 subdomains each with x, y dims (4, 4)
+    assert flattened.shape == (2, 16, 4)
+    # Check first subdomain of first sample has correct values
+    np.testing.assert_array_almost_equal(
+        flattened[0, :, 0],
+        np.array([0, 0, 0, 0, 0, 0, 1, 10, 0, 2, 3, 12, 0, 20, 21, 30]),
+    )
```
Comment (spencerkclark) [inline external/fv3fit/tests/reservoir/test_domain.py:204]: I know it's tedious, but it feels like you might as well test that all subdomains have the values you expect.

File: external/fv3fit/fv3fit/reservoir/model.py [hunk 4]
```diff
@@ -1,35 +1,43 @@
...
     _READOUT_SUBDIR = "readout"
     _METADATA_NAME = "metadata.yaml"
+    _SCALER_NAME = "scaler.npz"
+    _RANK_DIVIDER_NAME = "rank_divider.yaml"
+    _AUTOENCODER_SUBDIR = "autoencoder"
 
     def __init__(
         self,
+        input_variables: Iterable[Hashable],
+        output_variables: Iterable[Hashable],
```
Comment (spencerkclark) [inline external/fv3fit/fv3fit/reservoir/model.py:34]: Any reason you use `Iterable` here but `Sequence` in the `ReservoirTrainingConfig`?

**Review by spencerkclark — APPROVED** [review]
These changes generally look good to me.  Indeed I think things will be clearer when I take a closer look at #2179.

File: external/fv3fit/fv3fit/reservoir/domain.py [hunk 5]
```diff
@@ -150,11 +154,53 @@ def unstack_subdomain(self, tensor, with_overlap: bool):
                 f"{expected_stacked_size} (product of {unstacked_shape})."
             )
         unstacked_shape = (tensor.shape[0], *unstacked_shape)
+
         return np.reshape(tensor, unstacked_shape)
 
+    def flatten_subdomains_to_columns(self, data: tf.Tensor, with_overlap: bool):
+        # Divide into subdomains and flatten subdomains into columns.
+        # Dimensions [(time), x, y, feature_orig] -> [(time), feature_new, subdomain]
```
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/reservoir/domain.py:162]: In training there is a time dim, but during prediction there is not.

File: external/fv3fit/fv3fit/reservoir/model.py [hunk 6]
```diff
@@ -1,35 +1,43 @@
...
-    evens = v[::2]
-    odds = v[1::2]
-    c = np.empty((v.size,), dtype=v.dtype)
-    c[0::2] = evens ** 2
-    c[1::2] = odds
-    return c
+def _exists_in_dir(file_name, dir):
+    fs = fsspec.get_fs_token_paths(dir)[0]
+    contents = [os.path.basename(os.path.normpath(p)) for p in fs.ls(dir)]
+    return file_name in contents
```
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/reservoir/model.py:19]: It seems like it, thanks, I'll just use the more concise syntax.

File: external/fv3fit/fv3fit/reservoir/domain.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (spencerkclark) [reply external/fv3fit/fv3fit/reservoir/domain.py:162]: `stack_time_series_samples` (used by this method) assumes there is a time dimension though, no?

File: external/fv3fit/fv3fit/reservoir/model.py [hunk 7]
```diff
@@ -40,24 +48,37 @@ def __init__(
...
+        # prediction into denormalized physical values on cubedsphere coords.
+
         if self.square_half_hidden_state is True:
-            readout_input = _square_even_terms(self.reservoir.state, axis=0)
+            readout_input = square_even_terms(self.reservoir.state, axis=0)
         else:
             readout_input = self.reservoir.state
+        # For prediction over multiple subdomains (>1 column in reservoir state
+        # array), flatten state into 1D vector before predicting
+        if len(readout_input.shape) > 1:
```
[the remark this answers is not in the document]
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/reservoir/model.py:73]: Good point, it can be omitted.

File: external/fv3fit/fv3fit/reservoir/config.py (hunk 1 above)
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/reservoir/config.py:101]: No, updated

File: external/fv3fit/fv3fit/reservoir/config.py (hunk 2 above)
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/reservoir/config.py:65]: Yes, added.

File: external/fv3fit/fv3fit/reservoir/model.py (hunk 4 above)
Comment (AnnaKwa) [reply external/fv3fit/fv3fit/reservoir/model.py:28]: Iterable here matches the parent `Predictor` class. I updated the config class to also use Iterable.

**Pull request merged**